src/include/server/mir/scene/data_exchange.h

```diff
@@ -40,6 +41,24 @@ class DataExchangeSource
     /// Instructs the source to start sending it's data in the given mime type to the given fd.
     virtual void initiate_send(std::string const& mime_type, Fd const& target_fd) = 0;
 
+    /// The destination has cancelled the (DnD) transfer
+    virtual void cancelled() {}
+
+    /// The user has dropped the (DnD) transfer
+    virtual void dnd_drop_performed() {}
+
+    /// The DnD actions supported
+    virtual auto actions() -> uint32_t { return 0; }
+
+    /// target accepted a mime type
+    virtual void offer_accepted(std::optional<std::string> const& mime_type) {(void)mime_type; }
+
+    /// target indicated an action
+    virtual uint32_t offer_set_actions(uint32_t dnd_actions, uint32_t preferred_action)
+        { (void)dnd_actions; (void)preferred_action; return 0; }
+
+    /// The client has finished the (DnD) transfer
+    virtual void dnd_finished() {}
```

Comment:
Are these deliberately concrete implementations in this otherwise abstract interface? It's not a leftover from the spike?

Reply:
A bit of both. It avoids churn in the test code, which was convenient for the spike. As I've not decided on a test strategy I've not addressed it. Yet